fastcrypto/src/vrf.rs

```diff
@@ -0,0 +1,263 @@
+    use crate::error::FastCryptoError;
+    use crate::groups::ristretto255::{RistrettoPoint, RistrettoScalar, COFACTOR};
+    use crate::groups::{GroupElement, Scalar};
+    use crate::hash::{HashFunction, Sha512};
+    use crate::serde_helpers::ToFromByteArray;
+    use crate::traits::AllowedRng;
+    use crate::vrf::{VRFKeyPair, VRFPrivateKey, VRFProof, VRFPublicKey};
+    use elliptic_curve::hash2curve::{ExpandMsg, Expander};
+    use serde::{Deserialize, Serialize};
+
+    /// draft-irtf-cfrg-vrf-15 specifies suites for suite-strings 0x00-0x04 and notes that future
+    /// designs should specify a different suite_string constant, so we use "sui_vrf" here.
+    const SUITE_STRING: &[u8; 7] = b"sui_vrf";
+
+    /// Length of challenges. Must not exceed the length of field elements which is 32 in this case.
+    /// We set C_LEN = 16 which is the same as the existing ECVRF suites in draft-irtf-cfrg-vrf-15.
+    const C_LEN: usize = 16;
+
+    /// Default hash function
+    type H = Sha512;
+
+    /// Domain separation tag used in ecvrf_encode_to_curve (see also draft-irtf-cfrg-hash-to-curve-16)
+    const DST: &[u8; 49] = b"ECVRF_ristretto255_XMD:SHA-512_R255MAP_RO_sui_vrf";
+
+    #[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+    pub struct ECVRFPublicKey(RistrettoPoint);
+
+    impl VRFPublicKey for ECVRFPublicKey {
+        type PrivateKey = ECVRFPrivateKey;
+    }
+
+    impl ECVRFPublicKey {
+        /// Encode the given binary string as curve point. See section 5.4.1.2 of draft-irtf-cfrg-vrf-15.
+        fn ecvrf_encode_to_curve(&self, alpha_string: &[u8]) -> RistrettoPoint {
+            // This follows section 5.4.1.2 of draft-irtf-cfrg-vrf-15 for the ristretto255 group using
+            // SHA-512. The hash-to-curve for ristretto255 follows appendix B of draft-irtf-cfrg-hash-to-curve-16.
+
+            // Compute expand_message_xmd for the given message. Note that expand_message only returns
+            // and error if the len_in_bytes and output size of the hash function is out of bounds
+            // (https://github.com/mikelodder7/hash2field/blob/cdf56a2b722aeae25b8019945afe4cccec132f25/src/expand_msg_xmd.rs#L21),
+            // so we can safely unwrap since they are constants here.
+            let mut expanded_message =
+                elliptic_curve::hash2curve::ExpandMsgXmd::<sha2::Sha512>::expand_message(
+                    &[&self.0.compress(), alpha_string],
+                    DST,
+                    H::OUTPUT_SIZE,
+                )
+                .unwrap();
+
+            let mut bytes = [0u8; H::OUTPUT_SIZE];
+            expanded_message.fill_bytes(&mut bytes);
+            RistrettoPoint::map_to_point::<H>(&bytes)
+        }
+
+        /// Implements ECVRF_validate_key which checks the validity of a public key. See section 5.4.5
+        /// of draft-irtf-cfrg-vrf-15.
+        fn valid(&self) -> bool {
+            self.0 * RistrettoScalar::from(COFACTOR) != RistrettoPoint::zero()
```

Comment:
Do we need this for Ristretto? (actually all cofactor related code here)

Reply:
No, you're right! The ristretto group already factors out the cofactor, so we just need to check that it's not the identity.